python/external_models/oai_runner.py

```diff
@@ -0,0 +1,89 @@
+import torch
+import numpy as np
+from loguru import logger
+from typing import List, Tuple
+from abc import ABC, abstractmethod
+from transformers import (
+    AutoModelForCausalLM,
+    AutoModelForSeq2SeqLM,
+    AutoTokenizer,
+    AutoModelForTextEncoding,
+)
+import os
+import numpy as np
+import openai
+from openai import OpenAI
+from .external_parser import *
+
+
+class OpenAIRunner(Generator, Transformer):
+    client = OpenAI(
+        api_key=os.getenv("OPENAI_API_KEY"),
+    )
+
+    def __init__(self, **args):
+        self.client_kwargs: dict[str | str] = {
+            "model": args['model'],
+            "temperature": args['temperature'],
+            "max_tokens": args['max_tokens'],
+            "top_p": args['top_p'],
+            "frequency_penalty": 0,
+            "presence_penalty": 0,
+            "n": args['num_return_sequences'],
+            "timeout": args['openai_timeout'],
+            # "stop": args.stop, --> stop is only used for base models currently
+        }
+        self.name = self.client_kwargs["model"]
+
+    def generate(self, input: str, target_prefix: str = "") -> List[Tuple[str, float]]:
+        prompt = pre_process_input(self.name, input + target_prefix)
+        prompt = [
+            {"role": "user", "content": f"{prompt}"},
+
+        ]
+        try:
+            response = OpenAIRunner.client.chat.completions.create(
+                messages=prompt,
+                logprobs=True,
+                **self.client_kwargs,
+            )
+        except (
+            openai.APIError,
+            openai.RateLimitError,
+            openai.InternalServerError,
+            openai.OpenAIError,
+            openai.APIStatusError,
+            openai.APITimeoutError,
+            openai.InternalServerError,
+            openai.APIConnectionError,
+        ) as e:
+            print("Exception: ", repr(e))
+            print("Sleeping for 30 seconds...")
+            print("Consider reducing the number of parallel processes.")
+            # sleep(30)
```

Comment:
Why is the `sleep(30)` commented out here? Do we sleep for 30 seconds or do we not?

Reply:
this is to avoid connection error, I didn't have during testing but not sure if other cases.

Reply:
So we should have the `sleep(30)` here uncommented?

Reply:
This sleeping was deleted for faster generation.